### ISSUE TITLE
7$: zsh/bash tab completion for CLI

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#188]
- [x] macOS: Homebrew
  - [ ] ~~bash~~ -- **broken** (not sourced; crashes)
  - [x] fish
  - [x] zsh (Homebrew)
  - [x] zsh (Apple)
- [ ] macOS: manual install
  - [ ] zsh (Apple) -- **broken** (not sourced)
- [ ] Debian: manual install
  - [ ] bash
  - [ ] fish
  - [ ] zsh
- [x] Debian: .deb install
  - [x] bash
  - [x] fish
  - [x] zsh
- [ ] Debian: Nix install
  - [x] bash
  - [x] fish
  - [ ] zsh
- [x] Arch: manual install
  - [x] bash
  - [x] fish
  - [x] zsh

**Comment by maniac-en** [comment]
I claim this for-hire task. I do not expect payment after I complete this task. I will email the quick-lint-js team if I am assigned this task.

**Comment by strager** [comment]
We need to check that completion works when installing in the following ways:

- [ ] Homebrew
- [ ] Nix
- [ ] Debian/apt

Once everything is confirmed working, I think this task is bon.

**Comment by strager** [comment]
Homebrew on macOS: zsh, bash, and fish completion all **do not work** (pressing tab after `quick-lint-js -` completes nothing and shows no menu).

The completion files do seem to be installed:
* /usr/local/share/bash-completion/completions/quick-lint-js.bash
* /usr/local/share/fish/vendor_completions.d/quick-lint-js.fish
* /usr/local/share/zsh/site-functions/_quick-lint-js

I'm not using Homebrew-installed shells. Maybe I need to install the shells with Homebrew.

**Comment by strager** [comment]
If I use Homebrew-installed shells, things work a bit better:
* zsh completion works
* bash completion **does not work**
* fish completion works

**Comment by strager** [comment]
When I uninstalled the Homebrew-installed bash, I saw this message:

```
Uninstalling /usr/local/Cellar/bash/5.1.8... (157 files, 9.7MB)

Warning: The following may be bash configuration files and have not been removed!
If desired, remove them manually with `rm -rf`:
  /usr/local/etc/bash_completion.d
```

/usr/local/etc/bash_completion.d is not where quick-lint-js installed completion files.

I copied quick-lint-js.bash into /usr/local/etc/bash_completion.d/, but completion in bash (Homebrew or macOS-provided) still didn't work.

**Comment by maniac-en** [comment]
What happens when you copy `quick-lint-js.bash` to `/usr/local/share/bash-completion/completions` for a homebrew installed bash?

**Comment by strager** [comment]
> What happens when you copy `quick-lint-js.bash` to `/usr/local/share/bash-completion/completions` for a homebrew installed bash?

It's already installed there:
```
$ ls -la /usr/local/share/bash-completion/completions/
total 0
drwxr-xr-x  6 strager  admin  192 May 13 02:28 .
drwxr-xr-x  3 strager  admin   96 May 13 02:28 ..
lrwxr-xr-x  1 strager  admin   68 May 13 02:28 cmake -> ../../../Cellar/cmake/3.20.2/share/bash-completion/completions/cmake
lrwxr-xr-x  1 strager  admin   68 May 13 02:28 cpack -> ../../../Cellar/cmake/3.20.2/share/bash-completion/completions/cpack
lrwxr-xr-x  1 strager  admin   68 May 13 02:28 ctest -> ../../../Cellar/cmake/3.20.2/share/bash-completion/completions/ctest
lrwxr-xr-x  1 strager  admin   95 May 13 02:28 quick-lint-js.bash -> ../../../Cellar/quick-lint-js/HEAD-caf3440/share/bash-completion/completions/quick-lint-js.bash

$ head /usr/local/share/bash-completion/completions/quick-lint-js.bash
# Copyright (C) 2020  Matthew Glazar
# See end of file for extended copyright information.

_quick-lint-js () {
        local cur opts
        _init_completion -n = || return

        opts='--help --version --lsp-server --exit-fail-on= --output-format= --vim-file-bufnr='

        case $cur in
```

**Comment by strager** [comment]
Debian package (after a patch):
* zsh completion **does not work**
* bash completion **works**
* fish completion **works**

Patch for Debian:
```diff
diff --git a/dist/debian/debian/quick-lint-js.install b/dist/debian/debian/quick-lint-js.install
index 380fb75a..75ec4743 100644
--- a/dist/debian/debian/quick-lint-js.install
+++ b/dist/debian/debian/quick-lint-js.install
@@ -1,2 +1,5 @@
 usr/bin usr
+usr/share/bash-completion usr/share
+usr/share/fish usr/share
 usr/share/man usr/share
+usr/share/zsh usr/share
```

**Comment by strager** [comment]
Nix package:

* zsh completion **does not work**
* bash completion works
* fish completion works

It looks like Nix' zsh doesn't look in ~/.nix-profile at all. I don't think we can reasonably work around this.

**Comment by strager** [comment]
Steps to reproduce part of the macOS bash issues on Linux:
```
$ wget 'https://ftp.gnu.org/gnu/bash/bash-3.2.tar.gz'
$ tar xf bash-3.2.tar.gz
$ cd bash-3.2
$ ./configure
$ make -j40
$ ./bash --login -i
bash$ . ../completions/quick-lint-js.bash
bash$
```
Then type `quick-lint-js -` then press tab. Here is what I see:
```
strager@strapurp:~/Projects/quicklint-js/bash-3.2$ quick-lint-js -bash: _init_completion: command not found
```

Applying this workaround made completion of options work: https://gitlab.kitware.com/cmake/cmake/-/blob/96011ab06d8839010c181d501c58c8e914f3ba91/Auxiliary/bash-completion/cmake#L6-13

However, completion of files is broken (not exactly the same as on macOS, but similar):
```
strager@strapurp:~/Projects/quicklint-js/bash-3.2$ quick-lint-js bash: _tilde: command not found
```

The ball is in your court to fix this @maniac-en.

**Comment by strager** [comment]
Commit 8c56d9140f064b442a07e5568bb1d9ba3aeed0f8 fixed completion on Debian for apt-install quick-lint-js.

**Comment by wgrr** [comment]
FYI completion is missing `--stdin` from 19cf27d56

**Comment by maniac-en** [comment]
> $ wget 'https://ftp.gnu.org/gnu/bash/bash-3.2.tar.gz'

[bash-completion](https://github.com/scop/bash-completion) doesn't support bash version < `4.2`
![bash3.2](https://i.imgur.com/h0GjMYy.png)

> Applying this workaround made completion of options work: https://gitlab.kitware.com/cmake/cmake/-/blob/96011ab06d8839010c181d501c58c8e914f3ba91/Auxiliary/bash-completion/cmake#L6-13

Even if add this patch, we won't be able to complete following as the required functions are not available/defined:
- Providing file completions: function required `_filedir`
- Delete space after `*=` completions: function required `compopt -o nospace`

As per my research, [this](https://radiant-brushlands-42789.herokuapp.com/itnext.io/programmable-completion-for-bash-on-macos-f81a0103080b) explains in detail why don't they work in `bash 3.2`

TL;DR
- macOS runs an outdated version of bash, and need to updated
- bash completion has evolved a lot since version 3.2 and therefore the new features needed a new completion script.
- brew install bash-completion@2 provides completion features for bash >= `4.2` (supporting all completion features)
- brew install bash-completion provides completion features for bash 4.0 or lower. (lacks lots of completion features)

`Bash 4.2` works properly! (without the patch)
![bash4.2](https://i.imgur.com/syYnK94.png)

**Comment by maniac-en** [comment]
> FYI completion is missing `--stdin` from [19cf27d](https://github.com/quick-lint/quick-lint-js/commit/19cf27d56ab45f76643e49d745ea9bf0ac7a02c9)

I'll send a patch to add this!

**Comment by strager** [comment]
> > $ wget 'https://ftp.gnu.org/gnu/bash/bash-3.2.tar.gz'
> 
> [bash-completion](https://github.com/scop/bash-completion) doesn't support bash version < `4.2`
> ![bash3.2](https://camo.githubusercontent.com/991d8523ad88123bae9394cc833aac4f3ee10654eb61d8c1c04b1e4c20994045/68747470733a2f2f692e696d6775722e636f6d2f6830476a4d59792e706e67)
> 
> > Applying this workaround made completion of options work: https://gitlab.kitware.com/cmake/cmake/-/blob/96011ab06d8839010c181d501c58c8e914f3ba91/Auxiliary/bash-completion/cmake#L6-13
> 
> Even if add this patch, we won't be able to complete following as the required functions are not available/defined:
> 
>     * Providing file completions: function required `_filedir`
> 
>     * Delete space after `*=` completions: function required `compopt -o nospace`

Okay. Let's not bother supporting Bash on macOS. I think the default shell for newer versions of macOS is zsh anyway. (I will need to test Apple's zsh.)

**Comment by maniac-en** [comment]
completions tested for `zsh (Apple)` (I got it done through a friend's mac running zsh **NOT** installed from brew, confirmed using `brew list`)

![zsh_apple](https://cdn.discordapp.com/attachments/682346624907083787/843122836713373716/unknown.png)

**Note**: Even though I followed build instructions from [here](https://github.com/quick-lint/quick-lint-js/blob/master/docs/BUILDING.md#macos-homebrew-llvm-and-ninja), completions and man pages, were not installed in the required paths! For testing the completions, I put it there manually for testing.

**Comment by strager** [comment]
> completions tested for `zsh (Apple)` (I got it done through a friend's mac running zsh **NOT** installed from brew, confirmed using `brew list`)
> 
> ![zsh_apple](https://camo.githubusercontent.com/9b848aa394fef724d100dd9460b5bf59aa785da9b10e0ecaca4d9730fc021de1/68747470733a2f2f63646e2e646973636f72646170702e636f6d2f6174746163686d656e74732f3638323334363632343930373038333738372f3834333132323833363731333337333731362f756e6b6e6f776e2e706e67)
> 
> **Note**: Even though I followed build instructions from [here](https://github.com/quick-lint/quick-lint-js/blob/master/docs/BUILDING.md#macos-homebrew-llvm-and-ninja), completions and man pages, were not installed in the required paths! For completions, I put it there manually for testing.

What about quick-lint-js installed via Homebrew?

**Comment by strager** [comment]
> What about quick-lint-js installed via Homebrew?

By this, I mean using `brew install` to install quick-lint-js, not installing LLVM with Homebrew. https://quick-lint-js.com/install/#homebrew

**Comment by maniac-en** [comment]
> * [ ]  Arch: manual install
>   
>   * [ ]  bash
>   * [ ]  fish
>   * [ ]  zsh

Works for all of them, tried manual install both from `Linux: ninja` and `Linux: make`

**Comment by maniac-en** [comment]
Works fine when installed using [this](https://quick-lint-js.com/install/#homebrew). Completions and man pages automatically get installed.
> What about quick-lint-js installed via Homebrew?

**Comment by strager** [comment]
I think this is done enough.
